### PR TITLE
IGNITE-18442 Sql. JOIN with DISTINCT FROM fails.

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItJoinTest.java [hunk 1]
```diff
@@ -784,34 +790,66 @@ public void testNatural(JoinType joinType) {
...
+        } finally {
+            sql("DROP TABLE IF EXISTS t11");
+            sql("DROP TABLE IF EXISTS t22");
+        }
+    }
 
-        private final String[] disabledRules;
+    private static Stream<Arguments> joinTypes() {
+        List<Arguments> types = Arrays.stream(JoinType.values())
+                        .flatMap(v -> Stream.of(Arguments.of(v, false), Arguments.of(v, true))).collect(Collectors.toList());
```
Comment (ygerzhedovich) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItJoinTest.java:851]: you can just return stream here without collect result to list.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 2]
```diff
@@ -197,7 +199,7 @@ public static boolean isIdentity(List<? extends RexNode> projects, RelDataType i
 
     /** Binary comparison operations. */
     private static final Set<SqlKind> BINARY_COMPARISON =
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:201]: why we can't use org.apache.calcite.sql.SqlKind#BINARY_COMPARISON ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 3]
```diff
@@ -361,46 +364,58 @@ public static List<SearchBounds> buildHashIndexConditions(
...
             RelOptCluster cluster,
             RexNode condition,
-            RelDataType rowType
+            RelDataType rowType,
+            @Nullable ImmutableBitSet requiredColumns
     ) {
         condition = RexUtil.toCnf(builder(cluster), condition);
 
-        Int2ObjectMap<List<RexCall>> fieldsToPredicates = mapPredicatesToFields(condition, cluster);
+        Map<Integer, List<RexCall>> fieldsToPredicates = mapPredicatesToFields(condition, cluster);
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:377]: why you replace it?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 4]
```diff
@@ -361,46 +364,58 @@ public static List<SearchBounds> buildHashIndexConditions(
...
+
+        List<RelDataType> types = RelOptUtil.getFieldTypeList(rowType);
+
+        Mappings.TargetMapping mapping = null;
+
+        if (requiredColumns != null) {
+            mapping = Commons.inverseTrimmingMapping(types.size(), requiredColumns);
+        }
+
+        for (int fldIdx : fieldsToPredicates.keySet()) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:393]: Looks like int2ObjectEntrySet() more preferable for the case

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 5]
```diff
@@ -361,46 +364,58 @@ public static List<SearchBounds> buildHashIndexConditions(
...
+            List<RexCall> collFldPreds = fieldsToPredicates.get(fldIdx);
 
-        for (List<RexCall> collFldPreds : fieldsToPredicates.values()) {
             if (nullOrEmpty(collFldPreds)) {
                 break;
             }
 
             for (RexCall pred : collFldPreds) {
-                if (pred.getOperator().kind != SqlKind.EQUALS) {
+                if (pred.getOperator().kind != SqlKind.EQUALS && pred.getOperator().kind != IS_NOT_DISTINCT_FROM) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:401]: let's introduce colletion for both values and add explanation why we need it.

Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:401]: don`t understand here, for hash search we can use only **equality** or **is not distinct** operands, plz explain your idea.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java (hunk 2 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:201]: cause planner generates correlates with IS NOT DISTINCT FROM form, check for example :
AggregatePlannerTest#expandDistinctAggregates

**Comment by zstan** [comment]
@ygerzhedovich thanks for review, partially fixed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java (hunk 5 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:400]: pred.isA

**Comment by zstan** [comment]
@ygerzhedovich all done.

**Comment by zstan** [comment]
@ygerzhedovich all done

**Pull request closed without merging**